pkg/controller/controller.go

```diff
@@ -0,0 +1,288 @@
+package controller
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/submariner-io/admiral/pkg/federate"
+	mcservice "github.com/submariner-io/lighthouse/pkg/apis/lighthouse.submariner.io/v1"
+	core_v1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+type remoteServiceMap map[string]*mcservice.MultiClusterService
+type remoteClustersMap map[string]*remoteCluster
+
+type LightHouseController struct {
+	//remoteServices is a map that holds the remote services that are discovered.
+	remoteServiceMap remoteServiceMap
+
+	// remoteClusters is a map of remote clusters, which have been discovered
+	remoteClusters remoteClustersMap
+
+	processingMutex *sync.Mutex
+
+	federator federate.Federator
+
+	// Indirection hook for unit tests to supply fake client sets
+	newClientset func(kubeConfig *rest.Config) (kubernetes.Interface, error)
+}
+
+type remoteCluster struct {
+	stopCh           chan struct{}
+	clusterID        string
+	serviceInformer  cache.SharedIndexInformer
+	queue            workqueue.RateLimitingInterface
+	remoteServiceMap remoteServiceMap
+	federator        federate.Federator
+	processingMutex  *sync.Mutex
+}
+
+func New(federator federate.Federator) *LightHouseController {
+	return &LightHouseController{
+		federator:        federator,
+		remoteServiceMap: make(remoteServiceMap),
+		remoteClusters:   make(remoteClustersMap),
+		processingMutex:  &sync.Mutex{},
+
+		newClientset: func(c *rest.Config) (kubernetes.Interface, error) {
+			return kubernetes.NewForConfig(c)
+		},
+	}
+}
+
+func (c *LightHouseController) Start() error {
+	err := c.federator.WatchClusters(c)
+	if err != nil {
+		return fmt.Errorf("Could not register cluster watch: %v", err)
```

Comment:
As recommended [here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings), error strings should not be capitalized